File: MCP服务器/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/rust/src/revocation.rs
use chrono::Utc;

use crate::error::{Error, ErrorCode};
use crate::types::revocation::{RevocationDocument, RevocationReason, RevokedKey};

/// Build an empty revocation document for a domain.
pub fn build_revocation_document(domain: &str) -> RevocationDocument {
    RevocationDocument {
        schemapin_version: "1.2".to_string(),
        domain: domain.to_string(),
        updated_at: Utc::now().to_rfc3339(),
        revoked_keys: vec![],
    }
}

/// Add a revoked key to the document.
pub fn add_revoked_key(doc: &mut RevocationDocument, fingerprint: &str, reason: RevocationReason) {
    doc.revoked_keys.push(RevokedKey {
        fingerprint: fingerprint.to_string(),
        revoked_at: Utc::now().to_rfc3339(),
        reason,
    });
    doc.updated_at = Utc::now().to_rfc3339();
}

/// Check if a key fingerprint is revoked in a standalone revocation document.
/// Returns `Ok(())` if the key is not revoked, or `Err` if it is.
pub fn check_revocation(doc: &RevocationDocument, fingerprint: &str) -> Result<(), Error> {
    if let Some(rk) = doc
        .revoked_keys
        .iter()
        .find(|rk| rk.fingerprint == fingerprint)
    {
        return Err(Error::Verification {
            code: ErrorCode::KeyRevoked,
            message: format!("Key {} revoked: {:?}", fingerprint, rk.reason),
        });
    }
    Ok(())
}

/// Combined revocation check: checks both the simple revoked_keys list (from
/// the well-known response) and a standalone revocation document.
pub fn check_revocation_combined(
    simple_revoked: &[String],
    revocation_doc: Option<&RevocationDocument>,
    fingerprint: &str,
) -> Result<(), Error> {
    // Check simple list first
    if simple_revoked.iter().any(|k| k == fingerprint) {
        return Err(Error::Verification {
            code: ErrorCode::KeyRevoked,
            message: format!("Key {} found in revoked_keys list", fingerprint),
        });
    }
    // Check standalone doc
    if let Some(doc) = revocation_doc {
        check_revocation(doc, fingerprint)?;
    }
    Ok(())
}

/// Fetch a revocation document from a URL.
#[cfg(feature = "fetch")]
pub async fn fetch_revocation_document(url: &str) -> Result<RevocationDocument, Error> {
    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::none())
        .build()
        .map_err(|e| Error::Revocation(e.to_string()))?;

    let resp = client
        .get(url)
        .send()
        .await
        .map_err(|e| Error::Revocation(format!("Failed to fetch {}: {}", url, e)))?;

    if !resp.status().is_success() {
        return Err(Error::Revocation(format!(
            "HTTP {} fetching {}",
            resp.status(),
            url
        )));
    }

    let doc: RevocationDocument = resp
        .json()
        .await
        .map_err(|e| Error::Revocation(format!("Invalid JSON from {}: {}", url, e)))?;

    Ok(doc)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_revocation_document() {
        let doc = build_revocation_document("example.com");
        assert_eq!(doc.domain, "example.com");
        assert_eq!(doc.schemapin_version, "1.2");
        assert!(doc.revoked_keys.is_empty());
    }

    #[test]
    fn test_add_revoked_key() {
        let mut doc = build_revocation_document("example.com");
        add_revoked_key(&mut doc, "sha256:abc123", RevocationReason::KeyCompromise);
        assert_eq!(doc.revoked_keys.len(), 1);
        assert_eq!(doc.revoked_keys[0].fingerprint, "sha256:abc123");
        assert_eq!(doc.revoked_keys[0].reason, RevocationReason::KeyCompromise);
    }

    #[test]
    fn test_check_revocation_clean() {
        let doc = build_revocation_document("example.com");
        assert!(check_revocation(&doc, "sha256:abc123").is_ok());
    }

    #[test]
    fn test_check_revocation_revoked() {
        let mut doc = build_revocation_document("example.com");
        add_revoked_key(&mut doc, "sha256:abc123", RevocationReason::Superseded);
        assert!(check_revocation(&doc, "sha256:abc123").is_err());
        assert!(check_revocation(&doc, "sha256:other").is_ok());
    }

    #[test]
    fn test_check_revocation_combined_simple_list() {
        let simple = vec!["sha256:revoked1".to_string()];
        assert!(check_revocation_combined(&simple, None, "sha256:revoked1").is_err());
        assert!(check_revocation_combined(&simple, None, "sha256:clean").is_ok());
    }

    #[test]
    fn test_check_revocation_combined_doc() {
        let mut doc = build_revocation_document("example.com");
        add_revoked_key(&mut doc, "sha256:revoked2", RevocationReason::KeyCompromise);

        assert!(check_revocation_combined(&[], Some(&doc), "sha256:revoked2").is_err());
        assert!(check_revocation_combined(&[], Some(&doc), "sha256:clean").is_ok());
    }

    #[test]
    fn test_check_revocation_combined_both() {
        let simple = vec!["sha256:simple_revoked".to_string()];
        let mut doc = build_revocation_document("example.com");
        add_revoked_key(&mut doc, "sha256:doc_revoked", RevocationReason::Superseded);

        assert!(check_revocation_combined(&simple, Some(&doc), "sha256:simple_revoked").is_err());
        assert!(check_revocation_combined(&simple, Some(&doc), "sha256:doc_revoked").is_err());
        assert!(check_revocation_combined(&simple, Some(&doc), "sha256:clean").is_ok());
    }
}

File: reference-mcp-servers/security/ThirdKeyAI_SchemaPin_e5ce0c8_python/go/pkg/revocation/revocation.go
// Package revocation provides standalone revocation documents for SchemaPin v1.2.
package revocation

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// RevocationReason represents why a key was revoked.
type RevocationReason string

const (
	ReasonKeyCompromise         RevocationReason = "key_compromise"
	ReasonSuperseded            RevocationReason = "superseded"
	ReasonCessationOfOperation  RevocationReason = "cessation_of_operation"
	ReasonPrivilegeWithdrawn    RevocationReason = "privilege_withdrawn"
)

// RevokedKey represents a single revoked key entry.
type RevokedKey struct {
	Fingerprint string           `json:"fingerprint"`
	RevokedAt   string           `json:"revoked_at"`
	Reason      RevocationReason `json:"reason"`
}

// RevocationDocument represents a standalone revocation document.
type RevocationDocument struct {
	SchemapinVersion string       `json:"schemapin_version"`
	Domain           string       `json:"domain"`
	UpdatedAt        string       `json:"updated_at"`
	RevokedKeys      []RevokedKey `json:"revoked_keys"`
}

// BuildRevocationDocument creates an empty revocation document for a domain.
func BuildRevocationDocument(domain string) *RevocationDocument {
	return &RevocationDocument{
		SchemapinVersion: "1.2",
		Domain:           domain,
		UpdatedAt:        time.Now().UTC().Format(time.RFC3339),
		RevokedKeys:      []RevokedKey{},
	}
}

// AddRevokedKey adds a revoked key entry to the document.
func AddRevokedKey(doc *RevocationDocument, fingerprint string, reason RevocationReason) {
	now := time.Now().UTC().Format(time.RFC3339)
	doc.RevokedKeys = append(doc.RevokedKeys, RevokedKey{
		Fingerprint: fingerprint,
		RevokedAt:   now,
		Reason:      reason,
	})
	doc.UpdatedAt = now
}

// CheckRevocation checks if a fingerprint is revoked in the standalone document.
func CheckRevocation(doc *RevocationDocument, fingerprint string) error {
	for _, key := range doc.RevokedKeys {
		if key.Fingerprint == fingerprint {
			return fmt.Errorf("key %s is revoked: %s", fingerprint, key.Reason)
		}
	}
	return nil
}

// CheckRevocationCombined checks revocation against both simple list and standalone document.
func CheckRevocationCombined(simpleRevoked []string, doc *RevocationDocument, fingerprint string) error {
	for _, revoked := range simpleRevoked {
		if revoked == fingerprint {
			return fmt.Errorf("key %s is in simple revocation list", fingerprint)
		}
	}

	if doc != nil {
		return CheckRevocation(doc, fingerprint)
	}

	return nil
}

// FetchRevocationDocument fetches a standalone revocation document from a URL.
func FetchRevocationDocument(ctx context.Context, url string) (*RevocationDocument, error) {
	client := &http.Client{Timeout: 10 * time.Second}

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := client.Do(req) // #nosec G704 -- URL is from discovery document's revocation_endpoint
	if err != nil {
		return nil, fmt.Errorf("failed to fetch revocation document: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	var doc RevocationDocument
	if err := json.NewDecoder(resp.Body).Decode(&doc); err != nil {
		return nil, fmt.Errorf("failed to decode revocation document: %w", err)
	}

	return &doc, nil
}
